datasource/src/main/scala/quasar/physical/mongo/Mongo.scala

```diff
@@ -0,0 +1,143 @@
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import fs2.concurrent.{Queue, NoneTerminatedQueue}
+import fs2.Stream
+
+import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.physical.mongo.MongoResource.{Database, Collection}
+import quasar.Disposable
+
+import org.bson.{Document => _, _}
+import org.mongodb.scala._
+
+class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: MongoClient) {
+  import Mongo._
+
+  val F: ConcurrentEffect[F] = ConcurrentEffect[F]
+
+  def databases: Stream[F, Database] =
+    observableAsStream(client.listDatabaseNames).map(Database(_))
+
+  def databaseExists(database: Database): Stream[F, Boolean] =
+    databases.exists(_ === database)
+
+  def collections(database: Database): Stream[F, Collection] = for {
+    dbExists <- databaseExists(database)
+    res <- if (!dbExists) { Stream.empty } else {
+      observableAsStream(client.getDatabase(database.name).listCollectionNames()).map(Collection(database, _))
+    }
+  } yield res
+
+  def collectionExists(collection: Collection): Stream[F, Boolean] =
+    collections(collection.database).exists(_ === collection)
+
+  def findAll(collection: Collection): F[Stream[F, BsonValue]] = {
+    def getCollection(mongo: MongoClient) =
+      mongo.getDatabase(collection.database.name).getCollection(collection.name)
+
+    collectionExists(collection).compile.last.map(_ getOrElse false)
+      .flatMap(exists =>
+        if (exists) {
+          F.delay(observableAsStream(getCollection(client).find[BsonValue]()))
+        } else {
+          MonadResourceErr.raiseError(ResourceError.pathNotFound(collection.resourcePath))
+        })
+  }
+}
+
+object Mongo {
+
+  def observableAsStream[F[_]: ConcurrentEffect, A](obs: Observable[A]): Stream[F, A] = {
+    val F = ConcurrentEffect[F]
+
+    def run(action: F[Unit]): Unit = F.runAsync(action)(_ => IO.unit).unsafeRunSync
+
+    def handler(subVar: MVar[F, Subscription], cb: Option[Either[Throwable, F[A]]] => Unit): Unit = {
+      obs.subscribe(new Observer[A] {
+        override def onSubscribe(sub: Subscription): Unit = {
+          run(subVar.put(sub))
+          sub.request(1L)
+        }
+        override def onNext(result: A): Unit = {
+          cb(Some(Right(subVar.read.map(_.request(1L)).as(result))))
+        }
+        override def onError(e: Throwable): Unit = cb(Some(Left(e)))
+
+        override def onComplete(): Unit = cb(None)
+      })
+    }
+
+    def unsubscribe(s: Subscription): F[Unit] = F.delay {
+      if (!s.isUnsubscribed()) {
+        s.unsubscribe()
+      } else {
+        ()
+      }
+    }
+
+    def enqueueObservable(
+        obsQ: NoneTerminatedQueue[F, Either[Throwable, F[A]]],
+        subVar: MVar[F, Subscription])
+        : Stream[F, Unit] =
+      Stream.eval(F.delay(handler(subVar, { x => run(obsQ.enqueue1(x)) })))
+
+    (for {
+      obsQ <- Stream.eval(Queue.boundedNoneTerminated[F, Either[Throwable, F[A]]](32))
```

Comment:
We can revisit this later if we add batching support, but currently there won't ever be more than 1 item in the queue.